Submission logic for interop (EXPOSUREAPP-2737)

d4rken opened the pull request:
## Description
This adds submission logic for interop, adding `consentToFederation` and `visistedCountries` from the app config.

To successfully test our logic, additional refactoring was necessary to further rid ourselves of the singletons. This includes the verification, submission and playbook logic which are now separated and injected.

## How to test
* Perform the submission flow, and observe logcat.
* Review refactored tests and code.

## Draft TODO

- [x] Wait for #1213 to be merged to clean country access up here (currently hardcoded)
- [x] Wait for the PR for EXPOSUREAPP-2727 (#1246) to be created and merged due to overlap that could cause conflicts

d4rken committed:
Update protobuf format

d4rken committed:
Addition of contentToFederation and visistedCountries/supportedCountries.

TODO: Unit tests, pull supportedCountries from AppConfig (waiting for #1213)

d4rken committed:
Fix invalid package being assigned.
This would break daggers annotation processing as the class could not be resolved.

d4rken committed:
Improved existing and implemented new tests.

d4rken committed:
Fixed test regression due to refactoring.

d4rken committed:
Merge branch 'release/1.5.x' into feature/interop-keysubmission-logic

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/di/ApplicationComponent.kt
#	Corona-Warn-App/src/test/java/de/rki/coronawarnapp/service/applicationconfiguration/ApplicationConfigurationServiceTest.kt

d4rken committed:
Integrate pulling the supportedCountries data from the AppConfig for the submission transaction.
(Previously wasn't implemented yet, now it's merged)

mlenkeit commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/transaction/SubmitDiagnosisKeysTransaction.kt:
```diff
@@ -54,39 +54,42 @@ object SubmitDiagnosisKeysTransaction : Transaction() {
+        val visistedCountries = appConfigProvider.getAppConfig().supportedCountriesList
+
         executeState(RETRIEVE_TAN_AND_SUBMIT_KEYS) {
-            PlaybookImpl(WebRequestBuilder.getInstance()).submission(
-                registrationToken,
-                temporaryExposureKeyList
+            val submissionData = Playbook.SubmissionData(
+                registrationToken = registrationToken,
+                temporaryExposureKeys = temporaryExposureKeyList,
+                consentToFederation = true,
+                visistedCountries = visistedCountries
```
Question: Is the fallback to `DE` in case `visistedCountries` is empty built into the App Config?

d4rken replied:
It is not. Good point. Will fix.

d4rken commented on .idea/codeStyles/Project.xml:
```diff
@@ -3,7 +3,16 @@
       <option name="PACKAGES_TO_USE_STAR_IMPORTS">
         <value>
-          <package name="kotlinx.android.synthetic" withSubpackages="true" static="false" />
+          <package name="kotlinx.android.synthetic" alias="false" withSubpackages="true" />
```
Will fix this.

d4rken committed:
Merge branch 'release/1.5.x' into feature/interop-keysubmission-logic

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/http/ServiceFactory.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/di/ApplicationComponent.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/VerificationKeys.kt

d4rken committed:
Address merge conflicts, and a little refactoring, and a bit more tests.

d4rken committed:
Revert file change to Project.xml

d4rken committed:
AppConfigProvider should insert "DE" as fallback if the supportedCountry list is empty/not supplied.

chris-cwa commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/server/SubmissionServer.kt:
```diff
@@ -0,0 +1,85 @@
+        )
+    }
+
+    suspend fun submitKeysToServerFake() = withContext(Dispatchers.IO) {
```
why this method and not inject fake SubmissionServer when needed?

d4rken replied:
There is lot's of overlap between both methods, and they are both used in the in the playbook, we'd have to inject the real and the fake server, the fake one just uses different parameters.

I'm not opposed to injecting this somehow, just didn't see a clear way to do it when refactoring this. Would leave that for another PR, this one is already pretty huge.

chris-cwa commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/PaddingTool.kt:
```diff
@@ -0,0 +1,10 @@
+package de.rki.coronawarnapp.util
+
+object PaddingTool {
+    fun requestPadding(length: Int): String {
```
did I miss the unit test?

d4rken replied:
You did not, this needs one definitely one. Good point, will add :+1:

chris-cwa commented on Corona-Warn-App/src/test/java/testhelpers/extensions/JsonExtensions.kt:
```diff
@@ -0,0 +1,13 @@
+import com.google.gson.JsonObject
+import okhttp3.mockwebserver.MockResponse
+
+fun String.toComparableJson() = try {
+    Gson().fromJson(this, JsonObject::class.java).toString()
+} catch (e: Exception) {
+    throw IllegalArgumentException("'$this' wasn't valid JSON")
+}
```
how about
```suggestion
fun String.getValidatedJson(): String? = try {
    Gson().fromJson(this, JsonObject::class.java).toString()
} catch (e: Exception) {
    // log maybe
    null
}
```

d4rken replied:
This is an extension only used in tests, I'd prefer it to fail loudly. Will keep unless you have strong feelings about this.

mlenkeit commented on Corona-Warn-App/src/test/java/de/rki/coronawarnapp/appconfig/AppConfigProviderTest.kt:
```diff
@@ -185,6 +185,16 @@ class AppConfigServerTest : BaseIOTest() {
     }
 
+    @Test
+    fun `if supportedCountryList is empty, we insert DE as fallback`() {
+        coEvery { api.getApplicationConfiguration("DE") } returns APPCONFIG_BUNDLE.toResponseBody()
+        every { verificationKeys.hasInvalidSignature(any(), any()) } returns false
+
+        runBlocking {
+            createDownloadServer().getAppConfig().supportedCountriesList shouldBe listOf("DE")
+        }
+    }
+
```
@d4rken this fallback also applies then when the list of countries is displayed on the UI, isn't it? I think for the UI, there's a different kind of fallback. The `DE` fallback is only for the submission and the key download (i.e. non-UI coding).

d4rken replied:
I'll correct this. You are right, my approach would interfere with that behavior.

d4rken committed:
Don't insert the fallback country at application config provider level.
The UI requires the list to be empty, if it is empty in the config.
I've moved the fallback country insertion into the transaction.

d4rken committed:
Make detekt happy.

d4rken committed:
Make klint happy.

d4rken committed:
Merge branch 'release/1.5.x' into feature/interop-keysubmission-logic
